### PR TITLE
Improved bypass4netns implementation

### PR DESCRIPTION
**Pull request by naoki9911**
In this work, bypass4netns is entirely reimplemented and some new features added.

## New features
- Handling connections to bypassed sockets (`--handle-c2c-connections` option)
This enables containers to connect to bypassed (published) sockets.
- Policy-aware bypassing (`--tracer` option)
This enables bypass4netns to take care of iptables or other ACLs in intermediate NetNS.
- [Experimental] multi-node communication (`--mulitnode` option)
This enables multi-node communication without VXLAN. This feature is experimental and more work is required to apply in usernetes.
- Tests with existing applications (see GHA workflow)
Tests for above features with existing applications including statically linked golang binary.

## Removed features
- bypassing UDP sockets (SOCK_DGRAM)
Currently, this feature is not implemented yet.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks, but why was UDP removed?

**Comment by naoki9911** [comment]
> Thanks, but why was UDP removed?

New bypass4netns does not implement them **yet** for simplification of implementation.
Of course, bypass4netns should handle UDP sockets and I will add bypassing for UDP sockets.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -1,39 +1,215 @@
...
+        id: cache-restore
+        with:
+         key: lxc-image-base-${{ hashFiles('go.sum', 'test/init_test.sh') }}
+         path: /tmp/test-image.tar.zst
+         lookup-only: true
+
+      - name: setup lxd (v5.19)
+        id: s1
+        if: steps.cache-restore.outputs.cache-hit != 'true'
+        run: ./setup_lxd.sh
```
Comment (AkihiroSuda) [inline .github/workflows/test.yaml:47]: Can we move these scripts into a subdirectory?

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -1,39 +1,215 @@
...
+      - uses: actions/checkout@v4.1.1
+
+      - uses: actions/cache/restore@v3
+        id: cache-restore
+        with:
+         key: lxc-image-base-${{ hashFiles('go.sum', 'test/init_test.sh') }}
+         path: /tmp/test-image.tar.zst
+         lookup-only: true
+
+      - name: setup lxd (v5.19)
```
Comment (AkihiroSuda) [inline .github/workflows/test.yaml:44]: "v5.19" should be removed, as `setup_lxd.sh` does not pin the LXD version

File: .github/workflows/test.yaml (hunk 1 above)
Comment (naoki9911) [reply .github/workflows/test.yaml]: fixed in https://github.com/rootless-containers/bypass4netns/pull/39/commits/4b874fee803256d87c99a4765fd297743f3f5d5f

File: .github/workflows/test.yaml (hunk 2 above)
Comment (naoki9911) [reply .github/workflows/test.yaml]: fixed in https://github.com/rootless-containers/bypass4netns/pull/39/commits/4b874fee803256d87c99a4765fd297743f3f5d5f

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Comment by AkihiroSuda** [comment]
Tagged as https://github.com/rootless-containers/bypass4netns/releases/tag/v0.4.0

**Pull request merged**